L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc

```diff
@@ -0,0 +1,112 @@
+    if ((l1PFCand.id() == l1t::PFCandidate::ChargedHadron || l1PFCand.id() == l1t::PFCandidate::Electron) &&
+        std::abs(l1PFCand.eta()) < 2.5)
+      pfChargedHadrons_sort.push_back(l1PFCand);
+  std::sort(pfChargedHadrons_sort.begin(), pfChargedHadrons_sort.end(), [](l1t::PFCandidate i, l1t::PFCandidate j) {
+    return (i.pt() > j.pt());
+  });
+  auto lTaus = std::make_unique<l1t::PFTauCollection>();
+  if (pfChargedHadrons_sort.empty()) {
+    if (lTaus->empty()) {
+      PFTau dummy;
+      lTaus->push_back(dummy);
+    }
+    iEvent.put(std::move(lTaus), "L1PFTausNN");
+    return;
+  }
+  pfChargedHadrons_seeds.push_back(pfChargedHadrons_sort[0]);
+  for (unsigned int i0 = 1; i0 < pfChargedHadrons_sort.size(); i0++) {
+    bool pMatch = false;
+    for (unsigned int i1 = 0; i1 < pfChargedHadrons_seeds.size(); i1++) {
+      if (deltaR(pfChargedHadrons_seeds[i1], pfChargedHadrons_sort[i0]) < fConeSize_)
+        pMatch = true;
+    }
+    if (pMatch)
+      continue;
+    pfChargedHadrons_seeds.push_back(pfChargedHadrons_sort[i0]);
+    if (int(pfChargedHadrons_seeds.size()) > fMaxTaus_ - 1)
+      break;
+  }
+  for (unsigned int i0 = 0; i0 < pfChargedHadrons_seeds.size(); i0++) {
+    addTau(pfChargedHadrons_seeds[i0], (*l1PFCandidates), lTaus);
+  }
+  if (lTaus->empty()) {
+    PFTau dummy;
+    lTaus->push_back(dummy);
+  }
+  std::sort(lTaus->begin(), lTaus->end(), [](l1t::PFTau i, l1t::PFTau j) { return (i.pt() > j.pt()); });
+  iEvent.put(std::move(lTaus), "L1PFTausNN");
+}
+
+// create taus based on grid structure
+void L1NNTauProducer::addTau(const l1t::PFCandidate& iCand,
+                             const l1t::PFCandidateCollection& iParts,
+                             std::unique_ptr<l1t::PFTauCollection>& outputTaus) {
+  l1t::PFCandidateCollection pfTauCands;
+  TLorentzVector lTot;
+  lTot.SetPtEtaPhiM(0, 0, 0, 0);
+  TLorentzVector lCand;
+  lCand.SetPtEtaPhiM(0, 0, 0, 0);
+  int lId = 0;
+  for (const auto& l1PFCand : iParts) {
+    if (deltaR(iCand, l1PFCand) > fConeSize_)
+      continue;
+    TLorentzVector pVec;
+    pVec.SetPtEtaPhiM(l1PFCand.pt(), l1PFCand.eta(), l1PFCand.phi(), 0);
+    lTot += pVec;
+    if (deltaR(iCand, l1PFCand) < fTauSize_ &&
+        (l1PFCand.id() == l1t::PFCandidate::Electron || l1PFCand.id() == l1t::PFCandidate::ChargedHadron ||
+         l1PFCand.id() == l1t::PFCandidate::Photon)) {
+      lId++;
+      lCand += pVec;
+    }
+    pfTauCands.push_back(l1PFCand);
+  }
+  if (lTot.Pt() < fSeedPt_)
+    return;
+  std::sort(
+      pfTauCands.begin(), pfTauCands.end(), [](l1t::PFCandidate i, l1t::PFCandidate j) { return (i.pt() > j.pt()); });
+  float NN = fTauNNId_->compute(iCand, pfTauCands);
+  math::PtEtaPhiMLorentzVector tempP4(lCand.Pt(), lCand.Eta(), lCand.Phi(), lCand.M());
+  l1t::PFTau l1PFTau(tempP4, NN, 0, lId);
+  outputTaus->push_back(l1PFTau);
+}
+float L1NNTauProducer::deltaR(const l1t::PFCandidate& iPart1, const l1t::PFCandidate& iPart2) {
```

Comment:
use https://github.com/cms-sw/cmssw/blob/master/DataFormats/Math/interface/deltaR.h

Reply:
done.

Reply:
then this function should be deleted